### PR TITLE
Added zpp_bits - wonder if you can add it / check the results versus the other ones, I wasn't able to compile all of them

### PR DESCRIPTION
**Pull request by eyalz800**

### CODE REVIEW

**Comment by felixguendling** [comment]
Doesn't work for me:

```
[1/2] Building CXX object CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o
FAILED: CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o 
/usr/bin/c++ -DCISTA_FNV1A=1 -I/home/felix/Downloads/cpp-serialization-benchmark/include -I/home/felix/Downloads/cpp-serialization-benchmark/build -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/cista/include -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/flatbuffers/include -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/capnproto/c++/src -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/cereal/include -O3 -DNDEBUG -std=gnu++2a -MD -MT CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o -MF CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o.d -o CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o -c /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h: In function ‘constexpr ToType zpp::bits::std::bit_cast(const FromType&)’:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:42:37: error: expected primary-expression before ‘,’ token
   42 |     return __builtin_bit_cast(ToType, from);
      |                                     ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h: In member function ‘constexpr zpp::bits::errc zpp::bits::basic_out<ByteView>::serialize_one(auto:57&&)’:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:466:22: error: ‘memcpy’ is not a member of ‘zpp::bits::std’; did you mean ‘memcpy’?
  466 |                 std::memcpy(
      |                      ^~~~~~
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdint.h:26,
                 from /usr/lib/gcc/x86_64-linux-gnu/10/include/stdint.h:9,
                 from /home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:172,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:1:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:31:1: note: ‘memcpy’ declared here
   31 | __NTH (memcpy (void *__restrict __dest, const void *__restrict __src,
      | ^~~~~
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:501:22: error: ‘memcpy’ is not a member of ‘zpp::bits::std’; did you mean ‘memcpy’?
  501 |                 std::memcpy(m_data.data() + m_position,
      |                      ^~~~~~
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdint.h:26,
                 from /usr/lib/gcc/x86_64-linux-gnu/10/include/stdint.h:9,
                 from /home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:172,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:1:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:31:1: note: ‘memcpy’ declared here
   31 | __NTH (memcpy (void *__restrict __dest, const void *__restrict __src,
      | ^~~~~
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h: In member function ‘constexpr zpp::bits::errc zpp::bits::in<ByteView>::serialize_one(auto:72&&)’:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:759:22: error: ‘memcpy’ is not a member of ‘zpp::bits::std’; did you mean ‘memcpy’?
  759 |                 std::memcpy(
      |                      ^~~~~~
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdint.h:26,
                 from /usr/lib/gcc/x86_64-linux-gnu/10/include/stdint.h:9,
                 from /home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:172,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:1:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:31:1: note: ‘memcpy’ declared here
   31 | __NTH (memcpy (void *__restrict __dest, const void *__restrict __src,
      | ^~~~~
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:793:22: error: ‘memcpy’ is not a member of ‘zpp::bits::std’; did you mean ‘memcpy’?
  793 |                 std::memcpy(item.data(),
      |                      ^~~~~~
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdint.h:26,
                 from /usr/lib/gcc/x86_64-linux-gnu/10/include/stdint.h:9,
                 from /home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:172,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:1:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:31:1: note: ‘memcpy’ declared here
   31 | __NTH (memcpy (void *__restrict __dest, const void *__restrict __src,
      | ^~~~~
In file included from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h: At global scope:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:20:34: error: ‘members’ in namespace ‘zpp::bits’ does not name a template type
   20 |     using serialize = zpp::bits::members<3>;
      |                                  ^~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:24:34: error: ‘members’ in namespace ‘zpp::bits’ does not name a template type
   24 |     using serialize = zpp::bits::members<4>;
      |                                  ^~~~~~~
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5,
                 from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h: In instantiation of ‘constexpr auto zpp::bits::basic_out<ByteView>::serialize_many(auto:55&&, auto:56&& ...) [with auto:55 = std::vector<csb::zpp_bits_graph::node>&; auto:56 = {}; ByteView = std::vector<unsigned char>]’:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:389:30:   required from ‘constexpr auto zpp::bits::basic_out<ByteView>::operator()(auto:54&& ...) [with auto:54 = {std::vector<csb::zpp_bits_graph::node, std::allocator<csb::zpp_bits_graph::node> >&}; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:64:19:   required from ‘static constexpr auto csb::zpp_bits_graph::serialize(auto:95&, auto:96&) [with auto:95 = zpp::bits::basic_out<std::vector<unsigned char> >; auto:96 = csb::zpp_bits_graph]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:507:56:   required from ‘constexpr zpp::bits::errc zpp::bits::basic_out<ByteView>::serialize_one(auto:57&&) [with auto:57 = csb::zpp_bits_graph&; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:394:40:   required from ‘constexpr auto zpp::bits::basic_out<ByteView>::serialize_many(auto:55&&, auto:56&& ...) [with auto:55 = csb::zpp_bits_graph&; auto:56 = {}; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:648:41:   required from ‘constexpr auto zpp::bits::out<ByteView>::operator()(auto:66&& ...) [with auto:66 = {csb::zpp_bits_graph&}; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:76:64:   required from here
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:394:40: error: call of overloaded ‘serialize_one(std::vector<csb::zpp_bits_graph::node>&)’ is ambiguous
  394 |         if (auto result = serialize_one(first_item); failure(result))
      |                           ~~~~~~~~~~~~~^~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:444:20: note: candidate: ‘constexpr zpp::bits::errc zpp::bits::basic_out<ByteView>::serialize_one(auto:57&&) [with auto:57 = std::vector<csb::zpp_bits_graph::node>&; ByteView = std::vector<unsigned char>]’
  444 |     constexpr errc serialize_one(auto && item)
      |                    ^~~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:534:20: note: candidate: ‘constexpr zpp::bits::errc zpp::bits::basic_out<ByteView>::serialize_one(auto:59&&) [with SizeType = unsigned int; auto:59 = std::vector<csb::zpp_bits_graph::node>&; ByteView = std::vector<unsigned char>]’
  534 |     constexpr errc serialize_one(concepts::container auto && container)
      |                    ^~~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h: In instantiation of ‘constexpr auto zpp::bits::in<ByteView>::serialize_many(auto:70&&, auto:71&& ...) [with auto:70 = std::vector<csb::zpp_bits_graph::node>&; auto:71 = {}; ByteView = std::vector<unsigned char>]’:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:707:30:   required from ‘constexpr auto zpp::bits::in<ByteView>::operator()(auto:69&& ...) [with auto:69 = {std::vector<csb::zpp_bits_graph::node, std::allocator<csb::zpp_bits_graph::node> >&}; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:64:19:   required from ‘static constexpr auto csb::zpp_bits_graph::serialize(auto:95&, auto:96&) [with auto:95 = zpp::bits::in<std::vector<unsigned char> >; auto:96 = csb::zpp_bits_graph]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:799:56:   required from ‘constexpr zpp::bits::errc zpp::bits::in<ByteView>::serialize_one(auto:72&&) [with auto:72 = csb::zpp_bits_graph&; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:728:40:   required from ‘constexpr auto zpp::bits::in<ByteView>::serialize_many(auto:70&&, auto:71&& ...) [with auto:70 = csb::zpp_bits_graph&; auto:71 = {}; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:707:30:   required from ‘constexpr auto zpp::bits::in<ByteView>::operator()(auto:69&& ...) [with auto:69 = {csb::zpp_bits_graph&}; ByteView = std::vector<unsigned char>]’
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:78:47:   required from here
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:728:40: error: call of overloaded ‘serialize_one(std::vector<csb::zpp_bits_graph::node>&)’ is ambiguous
  728 |         if (auto result = serialize_one(first_item); failure(result))
      |                           ~~~~~~~~~~~~~^~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:741:20: note: candidate: ‘constexpr zpp::bits::errc zpp::bits::in<ByteView>::serialize_one(auto:72&&) [with auto:72 = std::vector<csb::zpp_bits_graph::node>&; ByteView = std::vector<unsigned char>]’
  741 |     constexpr errc serialize_one(auto && item)
      |                    ^~~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:827:5: note: candidate: ‘constexpr zpp::bits::errc zpp::bits::in<ByteView>::serialize_one(auto:74&&) [with SizeType = unsigned int; auto:74 = std::vector<csb::zpp_bits_graph::node>&; ByteView = std::vector<unsigned char>]’
  827 |     serialize_one(concepts::container auto && container)
      |     ^~~~~~~~~~~~~
In file included from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:1:
/home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:1263:7: error: redefinition of ‘int main(int, char**)’
 1263 |   int main(int argc, char** argv) {                                     \
      |       ^~~~
/home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:145:1: note: in expansion of macro ‘BENCHMARK_MAIN’
  145 | BENCHMARK_MAIN();
      | ^~~~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:1263:7: note: ‘int main(int, char**)’ previously defined here
 1263 |   int main(int argc, char** argv) {                                     \
      |       ^~~~
/home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:140:1: note: in expansion of macro ‘BENCHMARK_MAIN’
  140 | BENCHMARK_MAIN();BENCHMARK_TEMPLATE(BM_SERIALIZE, zpp_bits_bench)
      | ^~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```


> I wasn't able to compile all of them

Please only make an PR if it compiles at least for you.

**Comment by eyalz800** [comment]
What compiler are you using? @felixguendling your c++ library is missing std::bit_cast, can you try clang? I will try to also work around that in gcc if needed

**Comment by felixguendling** [comment]
Still no luck

```
[1/2] Building CXX object CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o
FAILED: CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o 
/usr/bin/clang++-12 -DCISTA_FNV1A=1 -I/home/felix/Downloads/cpp-serialization-benchmark/include -I/home/felix/Downloads/cpp-serialization-benchmark/build-clang -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/cista/include -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/flatbuffers/include -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/capnproto/c++/src -I/home/felix/Downloads/cpp-serialization-benchmark/external_lib/cereal/include -O3 -DNDEBUG -std=gnu++20 -MD -MT CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o -MF CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o.d -o CMakeFiles/cpp-serialization-benchmark.dir/src/main.cc.o -c /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc
In file included from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:20:34: error: no template named 'members' in namespace 'zpp::bits'
    using serialize = zpp::bits::members<3>;
                      ~~~~~~~~~~~^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:24:34: error: no template named 'members' in namespace 'zpp::bits'
    using serialize = zpp::bits::members<4>;
                      ~~~~~~~~~~~^
In file included from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:511:20: error: use of undeclared identifier 'serialize'
            return serialize(*this, item);
                   ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:567:35: note: in instantiation of function template specialization 'zpp::bits::basic_out<std::vector<unsigned char>>::serialize_one<csb::zpp_bits_graph::node &>' requested here
                if (auto result = serialize_one(item); failure(result))
                                  ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:395:27: note: in instantiation of function template specialization 'zpp::bits::basic_out<std::vector<unsigned char>>::serialize_one<unsigned int, std::vector<csb::zpp_bits_graph::node> &>' requested here
        if (auto result = serialize_one(first_item); failure(result))
                          ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:390:16: note: in instantiation of function template specialization 'zpp::bits::basic_out<std::vector<unsigned char>>::serialize_many<std::vector<csb::zpp_bits_graph::node> &>' requested here
        return serialize_many(items...);
               ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:64:12: note: in instantiation of function template specialization 'zpp::bits::basic_out<std::vector<unsigned char>>::operator()<std::vector<csb::zpp_bits_graph::node> &>' requested here
    return archive(self.nodes_);
           ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:508:47: note: in instantiation of function template specialization 'csb::zpp_bits_graph::serialize<zpp::bits::basic_out<std::vector<unsigned char>>, csb::zpp_bits_graph>' requested here
        } else if constexpr (requires { type::serialize(*this, item); }) {
                                              ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:508:41: note: in instantiation of requirement here
        } else if constexpr (requires { type::serialize(*this, item); }) {
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:395:27: note: in instantiation of function template specialization 'zpp::bits::basic_out<std::vector<unsigned char>>::serialize_one<csb::zpp_bits_graph &>' requested here
        if (auto result = serialize_one(first_item); failure(result))
                          ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:649:27: note: in instantiation of function template specialization 'zpp::bits::basic_out<std::vector<unsigned char>>::serialize_many<csb::zpp_bits_graph &>' requested here
            auto result = serialize_many(items...);
                          ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:76:49: note: in instantiation of function template specialization 'zpp::bits::out<std::vector<unsigned char>>::operator()<csb::zpp_bits_graph &>' requested here
  void serialize() { (void) zpp::bits::out{buf_}(zpp_bits_GRAPH); }
                                                ^
In file included from /home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:5:
In file included from /home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:5:
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:803:20: error: use of undeclared identifier 'serialize'
            return serialize(*this, item);
                   ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:868:35: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::serialize_one<csb::zpp_bits_graph::node &>' requested here
                if (auto result = serialize_one(item); failure(result))
                                  ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:729:27: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::serialize_one<unsigned int, std::vector<csb::zpp_bits_graph::node> &>' requested here
        if (auto result = serialize_one(first_item); failure(result))
                          ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:708:16: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::serialize_many<std::vector<csb::zpp_bits_graph::node> &>' requested here
        return serialize_many(items...);
               ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:64:12: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::operator()<std::vector<csb::zpp_bits_graph::node> &>' requested here
    return archive(self.nodes_);
           ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:800:47: note: in instantiation of function template specialization 'csb::zpp_bits_graph::serialize<zpp::bits::in<std::vector<unsigned char>>, csb::zpp_bits_graph>' requested here
        } else if constexpr (requires { type::serialize(*this, item); }) {
                                              ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:800:41: note: in instantiation of requirement here
        } else if constexpr (requires { type::serialize(*this, item); }) {
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:729:27: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::serialize_one<csb::zpp_bits_graph &>' requested here
        if (auto result = serialize_one(first_item); failure(result))
                          ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/../../../external_lib/zpp_bits/zpp_bits.h:708:16: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::serialize_many<csb::zpp_bits_graph &>' requested here
        return serialize_many(items...);
               ^
/home/felix/Downloads/cpp-serialization-benchmark/include/csb/benches/zpp_bits.h:78:33: note: in instantiation of function template specialization 'zpp::bits::in<std::vector<unsigned char>>::operator()<csb::zpp_bits_graph &>' requested here
      (void) zpp::bits::in{buf_}(deserialized_);
                                ^
/home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:145:1: error: redefinition of 'main'
BENCHMARK_MAIN();
^
/home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:1263:7: note: expanded from macro 'BENCHMARK_MAIN'
  int main(int argc, char** argv) {                                     \
      ^
/home/felix/Downloads/cpp-serialization-benchmark/src/main.cc:140:1: note: previous definition is here
BENCHMARK_MAIN();BENCHMARK_TEMPLATE(BM_SERIALIZE, zpp_bits_bench)
^
/home/felix/Downloads/cpp-serialization-benchmark/external_lib/google-benchmark/src/../include/benchmark/benchmark.h:1263:7: note: expanded from macro 'BENCHMARK_MAIN'
  int main(int argc, char** argv) {                                     \
      ^
5 errors generated.
ninja: build stopped: subcommand failed.
```

**Comment by eyalz800** [comment]
@felixguendling in your zpp_bits.h file do you have this line here? Just checking that you cloned the right version https://github.com/eyalz800/zpp_bits/blob/fe249864b3206472a7891ce72379acd2c932d2b4/zpp_bits.h#L326

**Comment by felixguendling** [comment]
You added the library with Git version https://github.com/eyalz800/zpp_bits/tree/dc8bdaccfc046d32b8da60ceb5bcef2b63835bdc - so no. Can you please create a PR that compiles and runs for you and post the results here? Thank you!

**Comment by eyalz800** [comment]
@felixguendling I am not super great on git submodules, I'll do it again from scratch

**Comment by eyalz800** [comment]
Confirmed it compiles on clang here, still figuring out g++, I forced push the submodule change.

**Comment by felixguendling** [comment]
I added CI so you can see what's working and what's not. I hope that helps.

Also please make sure to add benchmarks for every category.
Also make sure that you benchmark the same thing as the other benchmarks of the same category do.

**Comment by eyalz800** [comment]
@felixguendling I was able to install ubuntu and then fix many issues with g++, you need g++-11 at least. Waiting for the CI to run, thank you for that.

**Comment by eyalz800** [comment]
@felixguendling only issue I have is that both my mac and the ubuntu refuse to compile capnproto so I remove it manually any time I run the test.

**Comment by eyalz800** [comment]
Results from my ubuntu although it's single core and things may have ran in the background some I wouldn't trust them
```
╰─ ./cpp-serialization-benchmark                                                                                                                                                          ─╯






2021-12-17 23:44:41
Running ./cpp-serialization-benchmark
Run on (1 X 3200 MHz CPU )
Load Average: 0.52, 0.58, 0.59
***WARNING*** Library was built as DEBUG. Timings may be affected.
---------------------------------------------------------------------------------------------------------------------
Benchmark                                                          Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------------------
BM_SERIALIZE<cereal_bench>                                       422 ms          398 ms            2 size=37.829M
BM_SERIALIZE<fbs_bench>                                         5135 ms         5047 ms            1 size=62.998M
BM_SERIALIZE<cista_raw_bench>                                   3987 ms         3938 ms            1 size=176.378M
BM_SERIALIZE<cista_offset_bench>                                4397 ms         4328 ms            1 size=176.378M
BM_SAFE_DESERIALIZE<cista_offset_slim_bench>                    16.0 ms         15.3 ms           50
BM_SAFE_DESERIALIZE<fbs_bench>                                   511 ms          484 ms            2
BM_SAFE_DESERIALIZE<cista_offset_bench>                          927 ms          875 ms            1
BM_SAFE_DESERIALIZE<cista_raw_bench>                             987 ms          969 ms            1
BM_SAFE_DESERIALIZE<cereal_bench>                                407 ms          383 ms            2
BM_FAST_DESERIALIZE<cista_raw_bench>                             519 ms          516 ms            1
BM_TRAVERSE<cista_raw_bench>                                     632 ms          625 ms            1 nodeCount=1.87k
BM_TRAVERSE<cista_offset_slim_bench>                             723 ms          688 ms            1 nodeCount=1.87k
BM_TRAVERSE<cista_offset_bench>                                  716 ms          688 ms            1 nodeCount=1.87k
BM_TRAVERSE<cereal_bench>                                        675 ms          641 ms            1 nodeCount=1.87k
BM_TRAVERSE<fbs_bench>                                           774 ms          781 ms            1 nodeCount=1.87k
BM_FAST_DESERIALIZE_AND_TRAVERSE<cista_offset_slim_bench>        744 ms          719 ms            1 nodeCount=1.87k
BM_FAST_DESERIALIZE_AND_TRAVERSE<cista_offset_bench>            1185 ms          500 ms            1 nodeCount=1.87k
BM_FAST_DESERIALIZE_AND_TRAVERSE<cista_raw_bench>               1224 ms         1133 ms            2 nodeCount=1.87k
BM_FAST_DESERIALIZE_AND_TRAVERSE<fbs_bench>                      825 ms          812 ms            1 nodeCount=1.87k
BM_FAST_DESERIALIZE_AND_TRAVERSE<cereal_bench>                  1151 ms         1094 ms            1 nodeCount=1.87k
BM_SERIALIZE<zpp_bits_bench>                                     178 ms          177 ms            3 size=37.8066M
BM_SAFE_DESERIALIZE<zpp_bits_bench>                              180 ms          172 ms            4

```

**Comment by eyalz800** [comment]
@felixguendling it's waiting for the CI still, maybe you need to approve it.

The CI I forked is using gcc-10 but gcc-11 is the minimum where bit cast is supported I afraid

Link to CI run of the fork https://github.com/eyalz800/cpp-serialization-benchmark/runs/4565656299?check_suite_focus=true

File: src/main.cc [hunk 1]
```diff
@@ -136,4 +137,9 @@ BENCHMARK_TEMPLATE(BM_FAST_DESERIALIZE_AND_TRAVERSE, cereal_bench)
 BENCHMARK_TEMPLATE(BM_FAST_DESERIALIZE_AND_TRAVERSE, capnp_bench)
     ->Unit(benchmark::kMillisecond);
 
-BENCHMARK_MAIN();
\ No newline at end of file
+BENCHMARK_TEMPLATE(BM_SERIALIZE, zpp_bits_bench)
+    ->Unit(benchmark::kMillisecond);
+BENCHMARK_TEMPLATE(BM_SAFE_DESERIALIZE, zpp_bits_bench)
+    ->Unit(benchmark::kMillisecond);
```
Comment (felixguendling) [inline src/main.cc]: where's traverse + deserialize, etc.?

**Review by felixguendling — CHANGES_REQUESTED** [review]

File: src/main.cc (hunk 1 above)
Comment (eyalz800) [reply src/main.cc]: Added them in the appropriate places- In total there are 3 - serialize, deserialize, and traverse. Not sure if the deserialize and traverse together is necessary to benchmark in this case.

Comment (felixguendling) [reply src/main.cc]: As this is a very common use case (first deserialize, then traverse the data), I think this is a benchmark with high practical impact.

Comment (eyalz800) [reply src/main.cc]: Added it, it came out as a sum of deserialize and traverse separately

**Comment by eyalz800** [comment]
Hope this run will be successful, was quite an experience to figure everything out, has helped improve some stuff though https://github.com/eyalz800/cpp-serialization-benchmark/runs/4569343646?check_suite_focus=true

**Comment by eyalz800** [comment]
Hi thanks for this

**Comment by felixguendling** [comment]
You're welcome! I also updated the README file.

**Pull request merged**